src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs

```diff
@@ -14,43 +17,43 @@ namespace Microsoft.VisualStudio.LanguageServices.Experimentation
     [ExportWorkspaceService(typeof(IExperimentationService), ServiceLayer.Host), Shared]
     internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedObject, IExperimentationService
     {
-        private readonly object _experimentationServiceOpt;
-        private readonly MethodInfo _isCachedFlightEnabledInfo;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+
+        private Optional<Func<string, bool>> _isCachedFlightEnabled;
```

Comment:
since this is in master, i would think the risk woudl be lower for trying something like AsyncLazy (either roslyn's or JTFs).  I'd still prefer that myself.

Reply:
`AsyncLazy` includes a single-invocation guarantee for the initializer at the expense of complexity. This feature does not rely on this constraint.

Reply:
Well, complexity is in the eye of the beholder :)   When i see lazy intialization code that doesn't go through one of these helpers, then you have to then spend cycles going "ok... is this going to be ok in the presence of multiple initializers coming in".  Sometimes an abstraction/indirection can help toward reducing complexity.

That said, i'm not going to press on this.  This is all self contained, and if you feel that the current approach is fine, that's ok with me.  I might request though that we put a comment in the code explicitly stating that we've though about the case where this is called into from multiple threads, and it will still be safe in that case since it will just mean in rare cases we get a little extra computation that is then thrown away.   The comment at least makes it clear we thought about this though.  Thanks!

Reply:
@jasonmalinowski I think we need a helper for asynchronous optimistic initialization...

Basically the asynchronous form of `LazyInitializer.EnsureInitialized<T>(T, Func<T>)`.

Reply:
Yeah, I'd use AsyncLazy, if only because it means the field can be made readonly and nobody can access the other thing by accident. If this was perf-sensitive then maybe I'd care.

Reply:
p.s.: I should say "AsyncLazy"-like thing, as ours today probably isn't safe for JTF.